### DISCUSSION TITLE
Undo/Redo shortcuts don't work for theia extension custom editor widget

### DISCUSSION THREAD

**Discussion by Hanksha** [eclipse-theia/theia#12474] [Improvements]
This is a continuation of the discussion on this issue #9344 .

**Comment by vince-fugnitto** [comment]
Is there an issue with providing end-users the possibility to have more control over how and when to perform `undo` for your custom-view? In the following https://github.com/eclipse-theia/theia/issues/9344#issuecomment-1514938008 you mentioned that the monaco `undo` takes priority over your custom handler, so it seems you already have a handler present?

**Comment by Hanksha** [comment]
Yes I already have an handler registered for the "core.undo" command (and the "undo" command since I found out that the one called by the keybinding). The problem is that since the handler registered by monaco-command is always enabled (even when there are no text editors open) then my handler never gets called.

**Reply by vince-fugnitto** [reply]
As I mentioned, did you add a `when` clause to your handler to add priority over the others when your custom view is visible?

**Reply by Hanksha** [reply]
You mean a `when` clause to the keybinding? Because I don't see a `when` on the `CommandHandler` interface.

**Reply by vince-fugnitto** [reply]
> You mean a `when` clause to the keybinding? Because I don't see a `when` on the `CommandHandler` interface.

Yes, when registering the keybinding add a `when` clause that is enabled when your view is visible. It should take priority over monaco which does not have one. It is at least something to try.

**Reply by Hanksha** [reply]
Yes but that would mean registering a third undo and redo command (registering it for the same command won't help). I have a dozen custom editors across multiple extensions so the when clause would be pretty long.

From a design perspective you should have one command that represent a common action like undo, redo, copy, paste etc. and have one keybinding for it and multiple handlers only enabled correctly when they should. Then if you want a different keybinding for a specific editor you register another keybinding with a different "when" but for the same command.

I can try that workaround but I think it could be better to fix it in theia directly. I also found a workaround by filtering out "MonacoEditorCommandHandlers" via "FilterContribution" and registering my own implementation, that's likely to break though.

**Reply by vince-fugnitto** [reply]
As I mentioned before it's a little hard to help as I can't actually see the code, if you can provide a minimal example that is similar to what you're doing that would help. For example, what does your custom handler look like?

 In this case the monaco `undo` is correctly enabled (as it applies to focused or last focused editors) but you want to override it. A way forward can be to provide your own keybinding rule that applies to your custom-views (could be a base class you extend for all your custom-editors) only.

**Reply by Hanksha** [reply]
No it's **always** enabled, even without any text editors focused or opened, it returns true as a default, see https://github.com/eclipse-theia/theia/blob/master/packages/monaco/src/browser/monaco-command.ts#L179

**Reply by vince-fugnitto** [reply]
I'm not sure why by default it is `true` if you want to investigate a bit but I believe `undo` is very much tied to editors (https://github.com/eclipse-theia/theia/issues/2239). You may be interested in something that VS Code implemented https://github.com/microsoft/vscode/pull/98288/files if you don't want to add the workaround I mentioned.
